platforms/linux/hax_wrapper.c

```diff
@@ -0,0 +1,336 @@
+ *
+ *   1. Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "../../include/hax.h"
+#include "../../core/include/hax_core_interface.h"
+#include "../../core/include/ia32.h"
+
+#include <linux/atomic.h>
+#include <linux/mutex.h>
+#include <linux/slab.h>
+#include <linux/spinlock.h>
+#include <linux/spinlock_types.h>
+
+#include <asm/cmpxchg.h>
+
+int default_hax_log_level = 3;
+int max_cpus;
+hax_cpumap_t cpu_online_map;
+
+int hax_log_level(int level, const char *fmt,  ...)
+{
+    char buffer[256];
+    va_list args;
+
+    va_start(args, fmt);
+    vsnprintf(buffer, sizeof(buffer), fmt, args);
+    if (level >= default_hax_log_level) {
+        printk(KERN_ERR "haxm: %s", buffer);
+    }
+    va_end(args);
+    return 0;
+}
+
+uint32_t hax_cpuid(void)
+{
+    return smp_processor_id();
+}
+
+typedef struct smp_call_parameter {
+    void (*func)(void *);
+    void *param;
+    hax_cpumap_t *cpus;
+} smp_call_parameter;
+
+static void smp_cfunction(void *p)
+{
+    struct smp_call_parameter *info = p;
+    hax_cpumap_t *cpus;
+    uint32_t cpuid;
+
+    cpus = info->cpus;
+    cpuid = hax_cpuid();
+    if (*cpus & (0x1 << cpuid))
+        info->func(info->param);
+}
+
+int hax_smp_call_function(hax_cpumap_t *cpus, void (*scfunc)(void *), void * param)
+{
+    smp_call_parameter info;
+
+    info.func = scfunc;
+    info.param = param;
+    info.cpus = cpus;
+    on_each_cpu(smp_cfunction, &info, 1);
+    return 0;
+}
+
+/* XXX */
+int proc_event_pending(struct vcpu_t *vcpu)
+{
+    return vcpu_event_pending(vcpu);
+}
+
+void hax_disable_preemption(preempt_flag *eflags)
+{
+    mword flags;
+#ifdef __x86_64__
```

Comment:
Can we use `preempt_disable()` instead? Or you can add a TODO in case someone else wants to try it later.